rfc/rfc-66/rfc-66.md

```diff
@@ -0,0 +1,215 @@
+```
+
+### The Sorting Rules for Log Files from Different Writers
+
+The sorting rules is important because it somehow decides the natual order processing,
+especially when the event time field are the same, and we don't know which payload to choose when combining.
+Here we can keep using the log file name compactor of current codebase, that is:
+
+```sql
+order by delta_commit_time, version_number, write_token
+```
+
+For a single writer, its log files natual order is preserved by the auto increasing version_number,
+for multiple writers, we try the best to preserve the natual order still with the version number,
+but the write_token has deterministic priority in sorting, which breaks the file generation order.
+
+![log file sequence](log_file_sequence.png)
+
+### Format changes
+
+| Type                       | Changes                                                                                                                            |
+|----------------------------|------------------------------------------------------------------------------------------------------------------------------------|
+| Commit/compaction metadata | No changes                                                                                                                         |
+| Commit file name           | we are only going to add completion time in completed metadata filename for all actions                                            |
+| Log file format            | Adding deltacommit instant time to log file name. The file name does not contain base instant time; No changes to log block format |
+
+### Commit Protocol
+
+| Type                            | Content                                                                                                                                       |
+|---------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------|
+| Writer expectations             | Writer does not need to look up the latest file slice to fetch the base commit time anymore, it always uses the current instant time directly |
+| Conflict Resolution             | No explicit conflict resolution, it is handled by the compactor                                                                               |
+| Reader expectations             | Readers still query based on file slices, see the pseudocode for new file slicing                                                             |
+| Synchronization and correctness | Still need some mutex on commit log (timeline). Getting the completion time and adding to the timeline needs to happen atomically             |
+
+**Note:** Clustering and updates conflict resolution is not in the scope of non-blocking concurrency control, you still need the OCC for course-grained failures and retries. 
+
+## Appendix
+
+### Pseudocode for New File Slicing
```

Comment:
Just to make it concrete for the readers, we can add an example.
Let's take this example of one filegroup and consider the following sequence of commits:

1. `fg_t10` -> very first commit with start time, end time as `[t10, t20]`.
2. `l1` -> log file version 1 start time, end time as `[t21, t40]`.
3. `l2` -> concurrent log file version 2 `[t30, t50]`.
4. `fg_t60` -> base file due to compaction `[t60, t80]`.
5. `l3` -> concurrent log file version 3 `[t35, t90]`.

In this case, `file_slice_barriers` list is [t60, t10]. For a query at `t100`, `getAllFileSlices` should return the following list:
[
    {t60, fg_t60.parquet, {l3}},
    {t10, fg_t10.parquet, {l1, l2}}
]
This assumes that file slicing is done based on completion time.

Reply:
Yeah, I will add the example to the doc.